File: src/main/java/me/toidicakhia/xmlparser/XMLElement.java
package me.toidicakhia.xmlparser;

import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.Text;

import java.io.StringWriter;
import java.util.ArrayList;
import java.util.List;

import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

public class XMLElement {
    private final Element element;

    public XMLElement(Element element) {
        this.element = element;
    }

    public XMLElement getConfig(String key) {
        List<XMLElement> configs = getConfigList(key);
        return configs.isEmpty() ? null : configs.get(0);
    }

    public List<XMLElement> getConfigList(String key) {
        List<XMLElement> configs = new ArrayList<>();
        Node children = element.getFirstChild();

        while (children != null) {
            if (children instanceof Element) {
                Element childElement = (Element) children;
                if (childElement.getTagName().equalsIgnoreCase(key) && childElement.getParentNode() == element) {
                    configs.add(new XMLElement(childElement));
                }
            }
            children = children.getNextSibling();
        }
        return configs;
    }

    public List<String> getStringList(String key) {
        List<XMLElement> configs = getConfigList(key);
        List<String> strings = new ArrayList<>();

        for (XMLElement config : configs) {
            Node child = config.element.getFirstChild();
            while (child != null) {
                if (child instanceof Text) {
                    strings.add(child.getTextContent());
                    break;
                }
                child = child.getNextSibling();
            }
        }
        return strings;
    }

    public String getString(String key) {
        List<String> strings = getStringList(key);
        return strings.isEmpty() ? "" : strings.get(0);
    }

    public XMLElement addPropertyString(String key, String value) {
        Element child = element.getOwnerDocument().createElement(key);
        child.setTextContent(value);
        element.appendChild(child);
        return new XMLElement(child);
    }

    public void removePropertyWithValue(String key, String value) {
        Node child = element.getFirstChild();

        while (child != null) {
            if (child instanceof Element) {
                Element childElement = (Element) child;
                if (childElement.getTagName().equalsIgnoreCase(key) &&
                    childElement.getParentNode() == element &&
                    childElement.getTextContent().equals(value)) {
                    element.removeChild(child);
                    return;
                }
            }
            child = child.getNextSibling();
        }
    }

    public void editProperty(String key, String value, String newValue) {
        removePropertyWithValue(key, value);
        addPropertyString(key, newValue);
    }

    public void removeAllProperties(String key) {
        Node child = element.getFirstChild();

        while (child != null) {
            Node next = child.getNextSibling();
            if (child instanceof Element) {
                Element childElement = (Element) child;
                if (childElement.getTagName().equalsIgnoreCase(key) &&
                    childElement.getParentNode() == element) {
                    element.removeChild(child);
                }
            }
            child = next;
        }
    }

    public String toXML() {
        try {
            Transformer transformer = TransformerFactory.newInstance().newTransformer();
            transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
            transformer.setOutputProperty(OutputKeys.INDENT, "yes");
            transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4");

            StringWriter writer = new StringWriter();
            transformer.transform(new DOMSource(element), new StreamResult(writer));
            return writer.toString().trim();
        } catch (Exception e) {
            throw new RuntimeException("Error converting XML to String", e);
        }
    }
}

File: src/main/kotlin/me/toidicakhia/xmlparser/XMLElement.kt
package me.toidicakhia.xmlparser

import org.w3c.dom.Element
import org.w3c.dom.Text
import java.io.StringWriter
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

class XMLElement(val element: Element) {
	fun getConfig(key: String) = getConfigList(key).firstOrNull()

	fun getConfigList(key: String): List<XMLElement> {
		val children = element.childNodes
		val configs = mutableListOf<XMLElement>()

		for (i in 0..children.length) {
			val node = children.item(i) ?: continue

			if (node is Element && node.tagName.equals(key, true) && node.parentNode == element)
				configs.add(XMLElement(node))
		}

		return configs
	}

	fun getStringList(key: String): List<String> {
		val configs = getConfigList(key)
		val strings = mutableListOf<String>()

		for (config in configs) {
			val children = config.element.childNodes
			for (i in 0..children.length) {
				val node = children.item(i) ?: continue

				if (node is Text) {
					strings.add(node.data)
					break
				}
			}
		}

		return strings
	}

	fun getString(key: String) = getStringList(key).firstOrNull() ?: ""

	fun addPropertyString(key: String, value: String): XMLElement {
		val document = element.ownerDocument
		val child = document.createElement(key)
		child.textContent = value
		element.appendChild(child)

		return XMLElement(child)
	}

	fun removePropertyWithValue(key: String, value: String) {
		val children = element.childNodes

	    for (i in 0 until children.length) {
	        val node = children.item(i)
	        if (node is Element && node.parentNode == element && 
	        	node.tagName.equals(key, true) && node.textContent == value) {
	            element.removeChild(node)
	            return
	        }
	    }
	}

	fun editProperty(key: String, value: String, newValue: String) {
		removePropertyWithValue(key, value)
		addPropertyString(key, newValue)
	}

	fun removeAllProperties(key: String) {
		val children = element.childNodes

	    for (i in 0 until children.length) {
	        val node = children.item(i)
	        if (node is Element && node.parentNode == element && node.tagName.equals(key, true))
	            element.removeChild(node)
	    }
	}

	fun toXML(): String {
	    val transformer = TransformerFactory.newInstance().newTransformer()
	    transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes")
        transformer.setOutputProperty(OutputKeys.INDENT, "yes")
        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4")

	    val source = DOMSource(element)
	    val result = StringWriter()
	    transformer.transform(source, StreamResult(result))

	    return result.toString().trim()
	}
}